ls1intum/Artemis
# Feature/generalize submission process

### Checklist

### Motivation and Context


A lot of code duplication was introduced when adding new exercise type - File Upload. Resolves #819.
**NOTE**: This is a copy of PR #890 . Since I couldn't create a build on Bamboo for that PR, I created new branch and opened this PR.

### Description

- Created `GenericSubmissionResource` to define properties and methods that are used through modeling, text and file-upload submission resources.
- Created `GenericSubmissionRepository` which will be extended by concrete repositories. We have some shared queries that can be reused by passing `entityName`
- Moved some of the methods to `SubmissionService`
- Removed `Transactional` everywhere through the whole submission process.
- Switched to using primitive types whenever it is possible.
- Added more tests for different use cases of `TextSubmission`, `FileUploadSubmission` and `ModelingSubmission`.

**NOTE**: These changes are only related to the submission process, changes to the assessment process will be done in a separate PR.

### Steps for Testing


1. Log in to Artemis
2. Navigate to Overview
3. Select any course with open exercises
4. Submit File Upload/Modeling/Text exercise (you should be able to submit multiple times until the exercise due date). 
5. Start exercise after due date and do a late submission.
6. Go to Course Administration 
7. Open Tutor Course Dashboard
8. Check that the submissions are shown properly and tutor can assess them. Also check that participation has no submissions, results and student set.
9. Click 'Assess Next Submission' and check that you can see the next submission
